gpflow/experimental/check_shapes/bool_specs.py

```diff
@@ -0,0 +1,102 @@
+# Copyright 2022 The GPflow Contributors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Code for specifying and evaluating boolean expressions.
+"""
+from abc import ABC, abstractmethod
+from dataclasses import dataclass
+from typing import Any, Mapping, Tuple
+
+from .argument_ref import ArgumentRef
+from .error_contexts import ErrorContext, ObjectValueContext, ParallelContext, StackContext
+
+
+class ParsedBoolSpec(ABC):
+    """ A boolean expression. """
+
+    @abstractmethod
+    def get(self, arg_map: Mapping[str, Any], context: ErrorContext) -> Tuple[bool, ErrorContext]:
+        """ Evaluate this boolean value. """
+
+
+def _paren_repr(spec: ParsedBoolSpec) -> str:
```

Comment:
I'm not sure I want to do that. If I understand correctly you want to do:

```
a or b and c
```

instead of

```
a or (b and c)
```

? They might technically be the same, but I find the second option to much more readable, even if it's a bit longer.